### PR TITLE
Feature Branch: Updated Shopper Notices

### PR DESCRIPTION
**Pull request by mikejolley**
This is a _feature branch_ that introduces new components and styling for buyer notices. This includes standard banners, and snackbar notices (which share the same components). 

Once in a complete state, we'll reach out to other product teams to ensure this meets our needs internally, and potentially move (or plan to move) these components to a package for wider adoption.

Some brief background; aside from the new look (which we aim to roll out to all notices used on the frontend across WooCommerce), the main change is the move away from `@wordpress/components`. This is because the WordPress component library is intended to be used in the editor, not on the frontend of the store. Therefore we're creating new components specifically for that purpose and removing the dependency. Functionally, they are very similar.

**Todo:**
- [x] Create a readme/documentation for the new components
- [x] Add usage documentation to Storybook
- [x] Test light vs dark mode doesn't break styling
- [x] Extract styles and apply to other core notices (this may be separate to this PR)
- [x] Code review from Rubik
- [x] Functionality review from other product teams

## New Components

Note that there are new storybook entries for `NoticeBanner` and `SnackbarList` so feel free to review those there.

### `NoticeBanner`

Drop in replacement for the `Notice` component from `@wordpress/components` with custom icons and new styles.

[Docs](https://github.com/woocommerce/woocommerce-blocks/blob/313a259798a665a7329fd02e25cd129778e8fd4f/assets/js/base/components/notice-banner/README.md)

### `SnackbarList`

Accepts an array of notice objects and renders as snackbars in the bottom left corner of the screen. Uses `NoticeBanner` for the notices themselves, but has its own wrapper that handles the styling tweaks and timeout.

[Docs](https://github.com/woocommerce/woocommerce-blocks/blob/313a259798a665a7329fd02e25cd129778e8fd4f/assets/js/base/components/snackbar-list/README.md)

#### Accessibility

- [ ] I've tested using only a keyboard (no mouse)
- [ ] I've tested using a screen reader
- [ ] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [ ] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

#### Other Checks

- [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] This PR adds/removes an experimental interfaces and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] I tagged two reviewers because this PR makes queries to the database or I think it might have some security impact.

### Screenshots

This is from Storybook:
Screenshot 2023-03-07 at 17 09 19

### Testing

These new components can be seen in action in Storybook (run `npm run storybook` locally), or using the test cases below. 

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### User Facing Testing

These test cases show how to trigger the new notices during certain activities.

**Multiple packages notice**

1. Enable “Multiple Packages for WooCommerce” extension. 
2. Go to WooCommerce > Settings > Multiple packages 
3. Select "group by" product, then save changes.
4. Add several items to your cart.
5. Go to the cart page and choose “local pickup”
6. See the “Multiple shipments must have the same pickup location” notice

Screenshot 2023-03-07 at 15 16 11

**No Shipping Options**

1. Turn off multiple packages and disable all shipping rates
2. Go to checkout
3. See “There are no shipping options available. Please check your shipping address.” In the shipping options section.

Screenshot 2023-03-07 at 15 18 56

**No Payment Methods and top level errors**

1. Go to payment settings
2. Toggle of all payment methods
3. Go to checkout
4. See payment options section for the notice.
9. Try to place an order
10. See notice at the head of the checkout

Screenshot 2023-03-07 at 15 21 20
Screenshot 2023-03-07 at 15 35 23

<del>**Checkout Validation Errors**</del>

<del>1. During checkout for a UK address, enter a numeric postcode</del>
<del>2. See notice after server refresh</del>

Screenshot 2023-03-07 at 15 28 48

**Cart shipping calculator validation errors**

Repeat above test on the shipping calculator on the cart page.

Screenshot 2023-03-07 at 15 34 24

**Coupon form snackbars**

1. Go to the cart page
2. Apply a valid coupon. See snackbar. Snackbar should vanish on it’s own after 10000ms.
4. Remove the coupon. See snackbar.  Snackbar should vanish on it’s own after 10000ms.

Screenshot 2023-03-07 at 15 30 56
Screenshot 2023-03-07 at 15 30 19

### WooCommerce Visibility

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Dev Note

This update includes new designs/styles for store notices (errors, informational, success) as well as snack-bars, to give a uniform look and feel across all WooCommerce Blocks. 

In addition to this, if your store uses either the Cart Block or Checkout Block on the main cart and checkout pages, styles will also transfer to core notices. These new styles should take precedence over any existing theme styles due to the new markup and CSS being applied.

The basic HTML structure of the new notices is as follows:

```html
<div class="wc-block-components-notice-banner is-error">
     <svg />
     <div class="wc-block-components-notice-banner__content">MESSAGE</div>
</div>
```

There are different styles based on the notice status which include `error`, `success`, `info`, `default`, `warning`.

If you're not using the Cart or Checkout Blocks (if you're still using the shortcode based cart and checkout for example), existing notices rendered by WooCommerce will not be affected.

### Changelog

> New styles for error, info, and success notices across notices created by blocks. Additionally, existing notices in core receive new styles when Cart and Checkout Blocks are in use.

### CODE REVIEW

**Comment by mikejolley** [comment]
I have some tasks around testing but it would be great to get an early review from @woocommerce/rubik before I ask other teams to approve this work. I'll mark the PR as needing review, but please do not merge anything.

File: assets/js/base/components/snackbar-list/README.md
Comment (opr) [inline assets/js/base/components/snackbar-list/README.md]: Not sure if you want to say `To display snackbar notices` or `To display a snackbar notice`

File: assets/js/base/components/snackbar-list/README.md
Comment (opr) [inline assets/js/base/components/snackbar-list/README.md]: Might be worth mentioning that they disappear after a certain time too, I know you mentioned it further down but here would be good too.

File: assets/js/base/components/snackbar-list/README.md
Comment (opr) [inline assets/js/base/components/snackbar-list/README.md:61]: From the [@wordpess/notices documentation](https://developer.wordpress.org/block-editor/reference-guides/data/data-core-notices/#createnotice) - The Gutenberg equivalent is `explicitDismiss`.

`isDismissible` only works for regular notices and does not have any effect on snackbars.

With this in mind, do we want to keep ourselves in line with that, or do you think `isDismissible` is fine to keep in our snackbar list?

File: assets/js/base/components/snackbar-list/index.tsx
Comment (opr) [inline assets/js/base/components/snackbar-list/index.tsx]: TS error here, is it OK to pass an empty object instead of `undefined`?

File: assets/js/base/components/notice-banner/index.tsx
Comment (opr) [inline assets/js/base/components/notice-banner/index.tsx:30]: This is causing a TS error, because in https://github.com/woocommerce/woocommerce-blocks/blob/313a259798a665a7329fd02e25cd129778e8fd4f/assets/js/base/components/snackbar-list/index.tsx#L21,`notices` is a `Partial< NoticeType >[]` meaning its props are all considered optional.

File: assets/js/base/components/snackbar-list/index.tsx
Comment (opr) [inline assets/js/base/components/snackbar-list/index.tsx]: Could the type just be `NoticeType` here instead?

File: assets/js/base/components/notice-banner/utils.ts
Comment (opr) [inline assets/js/base/components/notice-banner/utils.ts:9]: Do we need a docblock here and for `getStatusIcon`?

File: assets/js/base/components/notice-banner/README.md
Comment (opr) [inline assets/js/base/components/notice-banner/README.md]:
```suggestion
Notices are color-coded to indicate the type of message being communicated, and also show an icon to reinforce the meaning of the message. The color and icon used for a notice are determined by the `status` prop.
```

File: assets/js/base/components/notice-banner/README.md
Comment (opr) [inline assets/js/base/components/notice-banner/README.md]:
```suggestion
The displayed message of a notice. Also used as the spoken message for assistive technology, unless `spokenMessage` is provided as an alternative message.
```

File: assets/js/base/components/notice-banner/README.md
Comment (opr) [inline assets/js/base/components/notice-banner/README.md]:
```suggestion
Optionally provided to change the spoken message for assistive technology. If not provided, the `children` prop will be used as the spoken message.
```

File: assets/js/base/components/notice-banner/README.md
Comment (opr) [inline assets/js/base/components/notice-banner/README.md]:
```suggestion
Status determines the color of the notice and the icon. Acceptable values are `success`, `error`, `info`, `warning`, and `default`.
```

File: assets/js/base/components/notice-banner/README.md
Comment (opr) [inline assets/js/base/components/notice-banner/README.md]:
```suggestion
In this example, the summary prop is used to indicate to the user that there are errors in the form submission. The list of error messages is rendered within the NoticeBanner component using an unordered list (`<ul>`) and list items (`<li>`). The `status` prop is set to `error` to indicate that the notice represents an error message.
```

File: assets/js/base/components/notice-banner/index.tsx
Comment (opr) [inline assets/js/base/components/notice-banner/index.tsx]: Maybe checking if `typeof event.preventDefault === 'function'` is better than just a truthiness check here?

File: assets/js/base/components/notice-banner/stories/index.tsx
Comment (opr) [inline assets/js/base/components/notice-banner/stories/index.tsx:59]: Noting that in PHPStorm there is a bug that shows an error here: `Element NoticeBanner doesn't have required attribute children `

If anyone else sees this, it's not an error but a [PHPStorm bug](https://youtrack.jetbrains.com/issue/WEB-48161)

File: assets/js/base/components/notice-banner/test/index.js
Comment (opr) [inline assets/js/base/components/notice-banner/test/index.js]: Should new files be `.tsx` ?

File: packages/checkout/components/store-notices-container/store-notices.tsx
Comment (opr) [inline packages/checkout/components/store-notices-container/store-notices.tsx]: There is an error when rendering `StoreNotice` because TS incorrectly assumes that `status` is a string, rather than one of the allowed strings.

To get around it, you could type `noticeProps` like:

```
Omit< NoticeBannerProps, 'children' > & {
  key: string;
}
```

**Review by opr — DISMISSED** [review]
Hey @mikejolley - I did an initial review and tested the functionality, here are some notes:

There are a few TS issues I mentioned inline.

**Snackbar List**
- Not sure if it's possible, but it would be great to add some actions to storybook to allow adding additional snackbar entries, and also make removing the snackbar items possible, to show off the full functionality.
- Tested in storybook and in the block, also tested with reduced motion, seems to be working OK.

**Notice Banner**
- Added a few comments for formatting the README, feel free to discard if you don't like the code formatting I suggested.
- Code for component looks good!
- Storybook looks good, though you can't add JSX elements as children via the controls, and when you load the `Error summary` element, the `children` prop is red and invalid, i don't think this is too serious though.
- image
- Tested in the blocks and found an issue when the server returns HTML
- image


**Store notices container**
- You didn't change this line, but line 70 in `packages/checkout/components/store-notices-container/index.tsx` complains because the `getNotices` action returns `Notice[]` but `additionalNotices` is `NoticeType[]` so trying to concat the two arrays results in an error, since they apparently don't intersect. There's an issue with the `status` type. Not sure on a fix for now, we could just ignore the error, but I don't feel like we should have to.
- Seems to be working fine, as `NoticeBanner` tested out (mostly) OK in the main notice area and the subcontexts.

File: assets/js/base/components/snackbar-list/README.md
Comment (mikejolley) [reply assets/js/base/components/snackbar-list/README.md:61]: I left this out intentionally. `explicitDismiss` changes the behaviour between click to dismiss, and click the X icon to dismiss. In our designs we only have variants with X close buttons, so we don't need this option.

Comment (opr) [reply assets/js/base/components/snackbar-list/README.md:61]: Gotcha, nice one!

**Comment by mikejolley** [comment]
I've taken care of that feedback I think @opr. I'll post about this on p2. Cheers.

**Comment by tarunvijwani** [comment]
@mikejolley Is this PR ready for review?

**Comment by mikejolley** [comment]
@tarunvijwani yes, I just rebased.

File: assets/js/base/components/snackbar-list/README.md
Comment (Aljullu) [inline assets/js/base/components/snackbar-list/README.md]: Heads-up that in RTL locales, it's displayed at the bottom right. Might be worth mentioning to avoid confusion?

File: assets/js/base/components/notice-banner/index.tsx
Comment (Aljullu) [inline assets/js/base/components/notice-banner/index.tsx:39]: I'm not sure to understand this sentence:

> WooCommerce blocks, themes, and plugins all use notices to indicate the result of an action, or to draw the user’s attention to necessary information.

I guess it comes from GB, but in the context of WC Blocks I feel it might generate confusion because other plugins and themes won't be using the same notice components as WC Blocks, right?

File: assets/js/base/components/notice-banner/style.scss
Comment (Aljullu) [inline assets/js/base/components/notice-banner/style.scss]: Do you know why these selectors are needed? In the examples, `.wc-block-components-notice-banner__content` doesn't seem to have a `div` inside by default.

imatge

File: assets/js/base/components/notice-banner/style.scss
Comment (Aljullu) [inline assets/js/base/components/notice-banner/style.scss:73]: Are there more buttons besides `.wc-block-components-notice-banner__dismiss`? If not, could we use that class name instead?

Also, I noticed that in TT3, with _Whisper_ styling, the button gets a weird border. Wondering if we should add some CSS resets to prevent themes messing with the style:

imatge

**Review by Aljullu — COMMENTED** [review]
I really like the new styles, good job! It also works great in light and dark themes. :100: 

I did some testing with the Mini Cart and Cart blocks and everything is working great. Leaving some notes below and inline: :point_down: 

* I see [`build/blocks-checkout.js` increased its size by 50%](https://github.com/woocommerce/woocommerce-blocks/pull/8659#issuecomment-1458546495). I guess there is not much that we can do, but wanted to mention just in case.
* We should add a changelog entry to the PR.
* It might also be good to add the `Mini Cart` label to the PR; as the Mini Cart can also display notices.

File: assets/js/base/components/notice-banner/style.scss
Comment (mikejolley) [reply assets/js/base/components/notice-banner/style.scss:73]: Some core notices do ("view more" links). I will add the reset though!

File: assets/js/base/components/notice-banner/style.scss
Comment (mikejolley) [reply assets/js/base/components/notice-banner/style.scss]: ah you're right this is no longer needed.

**Comment by ralucaStan** [comment]
@mikejolley it would be great to include in the docs a couple of images for each notification type so that it's easier for developers to make the connection between them

**Comment by mikejolley** [comment]
@ralucaStan Is that not the point of the storybook entries? There you can toggle between the various types.

**Comment by ralucaStan** [comment]
> @ralucaStan Is that not the point of the storybook entries? There you can toggle between the various types.

As a developer going through the docs, it's helpful to see the expected result without cloning the repo and starting Storybook. 
I recommend adding [this image](https://user-images.githubusercontent.com/90977/223496592-e461f7dc-acad-4b7b-b9a8-05f703ce1cb5.png) on the notices doc and the corresponding one for snack bars. You can also mention Storybook and if we have a public URL for our Storybook we could include that as well.

File: assets/js/base/components/notice-banner/style.scss
Comment (Aljullu) [inline assets/js/base/components/notice-banner/style.scss]: It hasn't been introduced in this PR, but we should add `:hover` and `:focus` styles for accessibility. Otherwise, there is no way to know if the focus is on the close button.

For reference, in the Mini Cart drawer close button we [play with the opacity](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/base/components/drawer/style.scss#L121-L134), maybe the same can be done here?

File: assets/js/base/components/notice-banner/style.scss
Comment (Aljullu) [inline assets/js/base/components/notice-banner/style.scss:46]: We should add `:hover` and `:focus` styles to this button as well.

**Review by Aljullu — APPROVED** [review]
Thanks for updating the PR, @mikejolley! I noticed that if a theme had custom styles for notices, they are not applied to these new notices. I guess that's per-design, but mentioning just in case:

Branch | Screenshot
--- | ---
`trunk` | imatge
This branch | imatge

I left a couple of inline comments below about buttons that need `:hover` styles, besides that, everything looks good to me, so pre-approving. (Heads-up that I didn't test https://github.com/woocommerce/woocommerce-blocks/pull/8732, which I see has been merged into this PR).

**Review by nefeline — APPROVED** [review]
Great Work 🎉 !

✅ Confirmed all notices are on Storybook
✅ Confirmed the notices are visible on cart and checkout pages

Minor inconsistency: when removing all items from cart, the style for that notice is slightly different:
Screenshot 2023-03-28 at 10 25 02

Might be related to what @Aljullu mentioned [over here](https://github.com/woocommerce/woocommerce-blocks/pull/8659#pullrequestreview-1361020139), but figured it would be worth flagging just in case.

**Comment by mikejolley** [comment]
Thanks @Aljullu and @nefeline. It seems some core notices are hardcoded and don't render using notice functions. I'm going to prepare a core patch to resolve that!

**Comment by mikejolley** [comment]
PR was created here https://github.com/woocommerce/woocommerce/pull/37514

**Pull request merged**
